core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowCompiler.scala

```diff
@@ -15,21 +15,105 @@ import edu.uci.ics.texera.web.workflowruntimestate.WorkflowAggregatedState.FAILE
 import edu.uci.ics.texera.web.workflowruntimestate.WorkflowFatalError
 import edu.uci.ics.texera.workflow.common.WorkflowContext
 import edu.uci.ics.texera.workflow.common.storage.OpResultStorage
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, Schema}
 import edu.uci.ics.texera.workflow.operators.sink.managed.ProgressiveSinkOpDesc
 import edu.uci.ics.texera.workflow.operators.visualization.VisualizationConstants
 
 import java.time.Instant
+import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
+case class WorkflowCompilationResult(
+    physicalPlan: PhysicalPlan,
+    operatorIdToInputSchemas: Map[OperatorIdentity, List[Option[Schema]]],
+    operatorIdToError: Map[OperatorIdentity, WorkflowFatalError]
+)
+
 class WorkflowCompiler(
     context: WorkflowContext
 ) extends LazyLogging {
 
+  /**
+    * Compile the workflow to logical plan and errors(if any)
+    * @param logicalPlanPojo the pojo parsed from workflow str provided by user
+    * @return LogicalPlan, and a Map from OpId to Op's error(this map is empty if there is no error)
+    */
+  def compileToLogicalPlan(
```

Comment:
I suggest we combine the two functions and provide a single `compile` API.

Reply:
done